src/systems/geometry.js

```diff
@@ -0,0 +1,140 @@
+var geometries = require('../core/geometry').geometries;
+var registerSystem = require('../core/system').registerSystem;
+var THREE = require('../lib/three');
+
+/**
+ * System for geometry component.
+ * Handle geometry caching.
+ *
+ * @member {object} cache - Mapping of stringified component data to Promises that resolve
+ *         THREE.Geometry objects.
+ * @member {object} cacheCount - Keep track of number of entities using a geometry to
+ *         know whether to dispose on removal.
+ */
+module.exports.System = registerSystem('geometry', {
+  init: function () {
+    this.cache = {};
+    this.cacheCount = {};
+  },
+
+  /**
+   * Reset cache. Mainly for testing.
+   */
+  clearCache: function () {
+    this.init();
+  },
+
+  /**
+   * Attempt to retrieve from cache.
+   *
+   * @returns {Object|null} A geometry if it exists, else null.
+   */
+  getOrCreateGeometry: function (data) {
+    var cache = this.cache;
+    var cacheCount = this.cacheCount;
+    var cachedGeometry;
+    var hash;
+
+    // Skip all caching logic.
+    if (data.skipCache) { return Promise.resolve(createGeometry(data)); }

```

Comment:
Mentioned in the description. In the scene with 4000 entities, each entity was looking up the cache, saw the cache was empty, and then started to create their own geometry object, resulting in 4000 geometry objects. I added the promises so when the first entity reaches the cache, it tells the rest that it's creating the object, and the rest of the entities wait.

Should we expect that if 4000 entities are initializing their geometry components, that they go one at a time serially?